### PR TITLE
feat: Automatically generate QDP plugins

### PR DESCRIPTION
**Pull request by bowang007**
# Description

This PR introduces a new feature which enables generating automatic plugin generation using TensorRT QDP feature. 

# Checklist:

- [ ] My code follows the style guidelines of this project (You can use the linters)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: examples/dynamo/auto_generate_plugin.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+# TensorRT QDP APIs
+torch_tensorrt.dynamo.conversion.plugins.generate_plugin("torchtrt_ex::elementwise_mul")
+
+
+# %%
+# Generating the Converter
+# -------------------------------------------------------------------
+# Given that we have defined the custom operator in PyTorch and TensorRT, we can now generate the converter for the operation.
+# As long as the namespace and names match, the following function will automatically generate the converter for the operation.
+torch_tensorrt.dynamo.conversion.plugins.generate_plugin_converter(
```
Comment (narendasan) [inline examples/dynamo/auto_generate_plugin.py:110]: Should these be unified? I like that these APIs exist separately but we should probably have another that does both

Comment (narendasan) [reply examples/dynamo/auto_generate_plugin.py:110]: So maybe something like 
- torch_tensorrt.dynamo.conversion.plugins.generate_plugin: (plugin)
- torch_tensorrt.dynamo.conversion.plugins.generate_plugin_converter: (converter for a QDP plugin)
- torch_tensorrt.dynamo.conversion.plugins.custom_op: (plugin + converter)

File: py/torch_tensorrt/dynamo/conversion/plugins/_generate_plugin.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
+import tensorrt_bindings.plugin as trtp
+from torch._dynamo.source import LocalSource
+import torch
+from torch.fx.experimental.symbolic_shapes import DimDynamic, ShapeEnv
+from sympy import lambdify
+from typing import Tuple
+from types import FunctionType
+from torch._subclasses.fake_tensor import FakeTensorMode
+from torch.fx.experimental.symbolic_shapes import ShapeEnv
+from sympy import lambdify
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/conversion/plugins/_generate_plugin.py:10]: Add sympy to pyproject.toml as a dependency

File: py/torch_tensorrt/dynamo/conversion/plugins/_generate_plugin.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+                impl_func_annotation[arg.name] = trtp.Tensor
+            elif arg.type.isSubtypeOf(torch._C.FloatType.get()):
+                register_func_annotation[arg.name] = float
+                impl_func_annotation[arg.name] = float
+            elif arg.type.isSubtypeOf(torch._C.IntType.get()):
+                register_func_annotation[arg.name] = int
+                impl_func_annotation[arg.name] = int
+            else:
+                raise ValueError("arg type is not handled")
+
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/conversion/plugins/_generate_plugin.py:56]: What other arg types would we need or is this it?

File: py/torch_tensorrt/dynamo/conversion/plugins/_generate_plugin.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+        register_func_annotation["return"] = Tuple[trtp.TensorDesc]
+        
+        impl_func_annotation["outputs"] = Tuple[trtp.Tensor]
+        impl_func_annotation["stream"] = int
+        
+        return args_input, kwargs_input, plugin_signature, plugin_impl_signagture, register_func_annotation, impl_func_annotation
+        
+    # Use the helper function to get the required signatures
+    args_input, kwargs_input, plugin_signature, plugin_impl_signature, register_func_annotation, impl_func_annotation = generate_signature(torch_op)
+    print(args_input)
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/conversion/plugins/_generate_plugin.py:84]: Make this debug info

File: py/torch_tensorrt/dynamo/conversion/plugins/_generate_plugin.py (hunk 3 above)
Comment (bowang007) [reply py/torch_tensorrt/dynamo/conversion/plugins/_generate_plugin.py:56]: @narendasan Other types are needed as well, this is just a temporary handling.

Comment (narendasan) [reply py/torch_tensorrt/dynamo/conversion/plugins/_generate_plugin.py:56]: Do you have a list?

File: py/torch_tensorrt/dynamo/conversion/plugins/_generate_plugin.py (hunk 4 above)
Comment (bowang007) [reply py/torch_tensorrt/dynamo/conversion/plugins/_generate_plugin.py:97]: @narendasan  when I use: 
```
    _LOGGER.debug(f"Plugin registration function: \n{codegen_plugin}")
```
It won't print anything. How to resolve this?

**Review by narendasan — COMMENTED** [review]
1. Figure out if QDP will hold the reference to the plugin for us
2. Add test cases for 1. (Tensor, Tensor) -> (Tensor) 2. (Tensor, int, float) -> (Tensor) 3. (Tensor, Tensor) -> (Tensor, Tensor)

**Comment by narendasan** [comment]
Rebase as well

File: examples/dynamo/auto_generate_plugin.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    x_vals = tl.load(X + offsets)
+    y_vals = tl.load(Y + offsets)
+    # Perform the element-wise multiplication
+    z_vals = x_vals * y_vals
+    # Store the result in Z
+    tl.store(Z + offsets, z_vals)
+
+
+@torch.library.custom_op("torchtrt_ex::elementwise_mul", mutates_args=())  # type: ignore[misc]
+def elementwise_mul(
```
Comment (narendasan) [inline examples/dynamo/auto_generate_plugin.py:67]: Small detail for the example, lets make use of the scalar inputs somehow and adjust the name of the op accordingly

File: py/torch_tensorrt/dynamo/conversion/plugins/_custom_op.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    generate_plugin_converter,
+)
+
+
+def custom_op(
+    op_name: str,
+    capability_validator: Optional[Callable[[Node, CompilationSettings], bool]] = None,
+    priority: ConverterPriority = ConverterPriority.STANDARD,
+    supports_dynamic_shapes: bool = False,
+):
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/conversion/plugins/_custom_op.py:17]: Add a docstring as this is a user API

File: py/torch_tensorrt/dynamo/conversion/plugins/_generate_plugin.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+            value,
+            source=source,
+            dynamic_dim=dynamic_dim,
+        ),
+        hint=value,
+        source=source,
+    )
+
+
+def generate_plugin(plugin_name: str):
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/conversion/plugins/_generate_plugin.py:27]: Add a docstring

**Review by narendasan — APPROVED** [review]
LGTM after docstrings are added

**Review by peri044 — COMMENTED** [review]
Added minor comments. LGTM

File: py/torch_tensorrt/dynamo/conversion/plugins/_generate_plugin.py [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+    plugin.__annotations__ = register_func_annotation
+
+    trtp.register(plugin_name)(plugin)
+
+    def _generic_plugin_impl(
+        outputs: Tuple[trtp.Tensor], stream: int, *args: Any, **kwargs: Any
+    ) -> None:
+        tensor_args = [elem for elem in args if isinstance(elem, trtp.Tensor)]
+        print(args)
+        non_tensor_args = [elem for elem in args if not isinstance(elem, trtp.Tensor)]
```
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/conversion/plugins/_generate_plugin.py:176]: Is it necessary to go for loops twice?

File: py/torch_tensorrt/dynamo/conversion/plugins/_generate_plugin_converter.py [hunk 9]
```diff
@@ -58,13 +57,24 @@ def custom_kernel_converter(
         # Assuming TensorRT preserves kwargs order like PyTorch does
         non_tensor_inputs = plugin.input_attrs
 
+        kwargs = {}
```
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/conversion/plugins/_generate_plugin_converter.py:60]: Do you intend to override `kwargs` here? If yes, it seems `kwargs` is not necessary in the arguments (line 46).

**Review by zewenli98 — COMMENTED** [review]
Added some minor comments

File: py/torch_tensorrt/dynamo/conversion/plugins/_generate_plugin.py [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+    # Function annotation is required for dynamic function to work in TensorRT.Plugin
+    plugin.__annotations__ = register_func_annotation
+
+    trtp.register(plugin_name)(plugin)
+
+    def _generic_plugin_impl(
+        outputs: Tuple[trtp.Tensor], stream: int, *args: Any, **kwargs: Any
+    ) -> None:
+        tensor_args = [elem for elem in args if isinstance(elem, trtp.Tensor)]
+        print(args)
```
[the remark this answers is not in the document]
Comment (bowang007) [reply py/torch_tensorrt/dynamo/conversion/plugins/_generate_plugin.py:175]: looks like I forgot to delete the debugging lines. Thanks!

File: py/torch_tensorrt/dynamo/conversion/plugins/_generate_plugin.py (hunk 8 above)
Comment (bowang007) [reply py/torch_tensorrt/dynamo/conversion/plugins/_generate_plugin.py:176]: I can write it as: 
```
tensor_args, non_tensor_args = [], []
for elem in args:
    (tensor_args if isinstance(elem, trtp.Tensor) else non_tensor_args).append(elem)
```
Since args won't be long I think the first one will be easier to understand?

File: py/torch_tensorrt/dynamo/conversion/plugins/_generate_plugin_converter.py (hunk 9 above)
Comment (bowang007) [reply py/torch_tensorrt/dynamo/conversion/plugins/_generate_plugin_converter.py:60]: yes, this kwargs cannot be erase since it's required in TorchTRT converter [here](https://github.com/pytorch/TensorRT/blob/2d3e06f54a1ae5a37219603e64410d9486ab469c/py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py#L855).

File: py/torch_tensorrt/dynamo/conversion/plugins/_custom_op.py [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    op_name: str,
+    capability_validator: Optional[Callable[[Node, CompilationSettings], bool]] = None,
+    priority: ConverterPriority = ConverterPriority.STANDARD,
+    supports_dynamic_shapes: bool = False,
+) -> None:
+    """
+    Generate the Plugin and corresponding Plugin Converter using external kernels and TensorRT Quick Deployable Plugin APIs.
+
+    Args:
+        plugin_name: the plugin name that is used to generate the plugin automatically.
```
[the remark this answers is not in the document]
Comment (zewenli98) [reply py/torch_tensorrt/dynamo/conversion/plugins/_custom_op.py:22]: it's a typo. args description doesn't match with args

File: py/torch_tensorrt/dynamo/conversion/plugins/_generate_plugin.py (hunk 8 above)
Comment (zewenli98) [reply py/torch_tensorrt/dynamo/conversion/plugins/_generate_plugin.py]: I think it's clearer and faster in this way:
```
tensor_args, non_tensor_args = [], []
for elem in args:
    if isinstance(elem, trtp.Tensor):
        tensor_args.append(elem)
    else:
        non_tensor_args.append(elem)
```
